testing/deployer/topology/compile.go

```diff
@@ -330,15 +404,78 @@ func compile(logger hclog.Logger, raw *Config, prev *Topology) (*Topology, error
 						foundPeerNames[c.Name][u.Peer] = struct{}{}
 					}
 
-					if u.ID.Name == "" {
-						return nil, fmt.Errorf("upstream service name is required")
-					}
 					addTenancy(u.ID.Partition, u.ID.Namespace)
+
+					if u.LocalAddress == "" {
+						// v1 defaults to 127.0.0.1 but v2 does not. Safe to do this generally though.
+						u.LocalAddress = "127.0.0.1"
+					}
+
+					if u.PortName != "" && n.IsV1() {
+						return nil, fmt.Errorf("explicit upstreams cannot use port names in v1")
+					}
+					if u.PortName == "" && n.IsV2() {
+						return nil, fmt.Errorf("explicit upstreams must use port names in v2")
+					}
 				}
 
 				if err := svc.Validate(); err != nil {
 					return nil, fmt.Errorf("cluster %q node %q service %q is not valid: %w", c.Name, n.Name, svc.ID.String(), err)
 				}
+
+				if n.IsV2() {
+					if implicitV2Services {
+						svc.V2Services = []string{svc.ID.Name}
+
+						var svcPorts []*pbcatalog.ServicePort
+						for name := range svc.Ports {
+							svcPorts = append(svcPorts, &pbcatalog.ServicePort{
+								TargetPort: name,
+								Protocol:   pbcatalog.Protocol_PROTOCOL_TCP, // TODO
```

Comment:
TODO:
For ergonomics I did not expose workload port protocols yet.

Reply:
This will be done in a future revision, as L7 isn't actually fully working yet at the envoy layer so configuring something other than TCP does not function.